worlds/mm2/__init__.py

```diff
@@ -0,0 +1,314 @@
+                weapons = list(map(lambda s: weapons_to_name[s], weaknesses))
+                valid_weapons = [item for item in progitempool
+                                 if item.name in weapons
+                                 and item.player == self.player]
+                placed_weapon = self.random.choice(valid_weapons)
+                weapon_name = next(name for name, idx in lookup_location_to_id.items()
+                                   if idx == 0x880101 + self.options.starting_robot_master.value)
+                weapon_location = self.get_location(weapon_name)
+                weapon_location.place_locked_item(placed_weapon)
+                progitempool.remove(placed_weapon)
+                fill_locations.remove(weapon_location)
+                self.multiworld.itempool.remove(placed_weapon)
+
+    def generate_output(self, output_directory: str) -> None:
+        try:
+            patch = MM2ProcedurePatch(player=self.player, player_name=self.multiworld.player_name[self.player])
+            patch_rom(self, patch)
+
+            self.rom_name = patch.name
+
+            patch.write(os.path.join(output_directory,
+                                     f"{self.multiworld.get_out_file_name_base(self.player)}{patch.patch_file_ending}"))
+        except Exception:
+            raise
+        finally:
+            self.rom_name_available_event.set()  # make sure threading continues and errors are collected
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        return {
+            "death_link": self.options.death_link.value,
+            "weapon_damage": self.weapon_damage
+        }
+
+    def interpret_slot_data(self, slot_data: Dict[str, Any]) -> Dict[str, Any]:
+        local_weapon = {int(key): value for key, value in slot_data["weapon_damage"].items()}
+        return {"weapon_damage": local_weapon}
+
+    def modify_multidata(self, multidata: Dict[str, Any]) -> None:
+        # wait for self.rom_name to be available.
+        self.rom_name_available_event.wait()
+        rom_name = getattr(self, "rom_name", None)
+        # we skip in case of error, so that the original error in the output thread is the one that gets raised
+        if rom_name:
+            new_name = base64.b64encode(bytes(self.rom_name)).decode()
+            multidata["connect_names"][new_name] = multidata["connect_names"][self.multiworld.player_name[self.player]]
+
+    def collect(self, state: "CollectionState", item: "Item") -> bool:
+        change = super().collect(state, item)
+        if change and item.name in self.item_name_groups["Weapons"]:
+            # need to evaluate robot master access
+            weapon = next((wp for wp in weapons_to_name if weapons_to_name[wp] == item.name), None)
+            if weapon:
+                for rbm in robot_masters:
+                    if self.weapon_damage[weapon][rbm] >= minimum_weakness_requirement[weapon]:
+                        state.prog_items[self.player][robot_masters[rbm]] = 1
```

Comment:
If I'm understanding this correctly, this really is just a static mapping of "weapons that can beat each robot master", right?

I feel like this might be overengineered & slow then, could you not build a `weapons_that_can_beat_each_robot_master: Dict[..., List[str]` and then use `state.has_any`?

Reply:
Re-raising this as part of my proper review

Reply:
FTR, this is what the collect replaced.

```
def can_defeat_enough_rbms(state: "CollectionState", player: int, required: int):
    can_defeat = 0
    for boss in range(8):
        if any(weapon_damage[weapon][boss] > 0 for weapon in weapon_damage):
            can_defeat += 1
    return can_defeat >= required
```
which with the amount of times rules get called, should perform worse than collect (especially since the collect code can filter down to relevant items only).